pyexchange/gemini.py

```diff
@@ -0,0 +1,215 @@
+#
+# Copyright (C) 2020 Exef
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+from logging import error
+
+from pyexchange.api import PyexAPI
+import hmac
+import requests
+import json
+import base64
+import hashlib
+import datetime, time
+
+from urllib.parse import urlencode
+
+from pymaker import Address, Wad
+from pymaker.util import http_response_summary
+from pyexchange.model import Order, Trade
+from typing import Optional, List
+
+
+class GeminiTrade(Trade):
+  @staticmethod
+  def from_my_trade(pair, item):
+    return GeminiTrade(trade_id=str(item["tid"]),
+                       timestamp=int(item["timestamp"]),
+                       pair=pair,
+                       is_sell=True if item["type"].lower() == "sell" else False,
+                       price=Wad.from_number(item["price"]),
+                       amount=Wad.from_number(item["amount"]))
+
+  @staticmethod
+  def from_trade(pair, item):
+    return GeminiTrade(trade_id=str(item["tid"]),
+                       timestamp=int(item["timestamp"]),
+                       pair=pair,
+                       is_sell=True if item["type"].lower() == "sell" else False,
+                       price=Wad.from_number(item["price"]),
+                       amount=Wad.from_number(item["amount"]))
+
+class GeminiOrder(Order):
+  @staticmethod
+  def create(item):
+    return GeminiOrder(order_id=item["order_id"],
+                       pair=item["symbol"].upper(),
+                       is_sell=True if item["side"].lower() == "sell" else False,
+                       price=Wad.from_number(item["price"]),
+                       amount=Wad.from_number(item["executed_amount"]),
+                       timestamp=int(item["timestamp"]))
+
+
+class GeminiApi(PyexAPI):
+  """ Gemini Exchange API interface
+
+  Implemented based on https://docs.gemini.com/rest-api/?python
+  """
+
+  logger = logging.getLogger()
+  
+  def __init__(self, api_server: str, api_key: str, api_secret: str, timeout: float) -> None:
+    assert(isinstance(api_server, str))
+    assert(isinstance(api_key, str))
+    assert(isinstance(api_secret, str))
+    assert(isinstance(timeout, float))
+
+    self.api_server = api_server
+    self.api_key = api_key
+    self.api_secret = api_secret
+    self.timeout = timeout
+
+  def get_balances(self):
```

Comment:
Is there a reason this returns a dictionary? `Market-Maker-Keeper` expects a `List` I believe. Also, adding the return type would be helpful for debugging.

Reply:
As far as I know, it's unspecified what this method should return. Usually, it returns an unchanged response from an API, and let's keeper handle it.

Reply:
I've actually had to convert it from `dict` into `List` before in DyDx pyexchange, as a `dict` is never expected. You can of course handle it in the consumer, but you will then need to make the identical conversion to a `list` in both `inventory-service`, and `market-maker-keeper`.